module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,2159 @@
+ *  c  |       ..group 5       |        group 6..      |       | row 4
+ *  e  +-----------+-----------+-----------------------+-------+
+ *  1  |..group 6  |            group 7                |       | row 5
+ *     +===+===+===+===+===+===+===+===+===+===+===+===+===+===+
+ *     | 3 | 5 | 10| 8 | 6 | 11| 12| 0 | 2 | 4 | 7 | 1 | 9 | 13| device map 2
+ *  s  +===+===+===+===+===+===+===+===+===+===+===+===+===+===+
+ *  l  |              group 8              |  group 9..|       | row 6
+ *  i  +-----------------------------------------------+-------|
+ *  c  |       ..group 9       |        group 10..     |       | row 7
+ *  e  +-----------------------+-----------------------+-------+
+ *  2  |..group 10 |            group 11               |       | row 8
+ *     +-----------+-----------------------------------+-------+
+ *
+ * This layout has several advantages over requiring that each row contain
+ * a whole number of groups.
+ *
+ * 1. The group count is not a relevant parameter when defining a dRAID
+ *    layout. Only the group width is needed, and *all* groups will have
+ *    the desired size.
+ *
+ * 2. All possible group widths (<= physical disk count) can be supported.
+ *
+ * 3. The logic within vdev_draid.c is simplified when the group width is
+ *    the same for all groups (although some of the logic around computing
+ *    permutation numbers and drive offsets is more complicated).
+ *
+ * N.B. The following array describes all valid dRAID permutation maps.
+ * Each row is used to generate a permutation map for a different number
+ * of children from a unique seed. The seeds were generated and carefully
+ * evaluated by the 'draid' utility in order to provide balanced mappings.
+ * In addition to the seed a checksum of the in-memory mapping is stored
+ * for verification.
+ *
+ * The commented value after the row is the average imbalance ratio. This
+ * ratio is a measure of how evenly balanced the resilver IO is over the
+ * children when there is one or more fault.  It is calculated by simulating
+ * all possible single and double disk faults for group widths up to 19
+ * devices (16d+3p).  A value of 1.0 indicates a perfectly balanced map;
+ * all children will perform the same number of IOs regardless of which
+ * device(s) failed. The average imbalance ratio provides a measure of
+ * how fair the mapping is for all possible dRAID configurations.
+ *
+ * Note these values are hard coded and must never be changed.  Existing
+ * pools depend on the same mapping always being generated in order to
+ * read and write from the correct locations.  Any change would make
+ * existing pools completely inaccessible.
+ */
+static draid_map_t draid_maps[VDEV_DRAID_MAX_MAPS] = {
+	{   2, 256, 0xd27b123486e72fe2, 0x000000003848433d, NULL }, /* 1.000 */
+	{   3, 256, 0x625f944e90fc7b1f, 0x00000000a8bfd5c4, NULL }, /* 1.000 */
+	{   4, 256, 0xc9ea9ec82340c885, 0x00000001819d7c69, NULL }, /* 1.000 */
```

Comment:
> It might be more clear to have this array indexed by the number of children, rather than having the number be a field in it?

I gave this a try and wasn't thrilled with the final result.  Indexing by the number of children means we have two unused rows at the start of the table and I felt it was overall less clear without the explicit child count.  I'd prefer to leave this as is.